### PR TITLE
Makes `fake_request` honor bytes in Python 3

### LINKED ISSUE #8: Python 3 compatibility
In file: `pytest_falcon/plugin.py`
```
>       resp.body = b''.join(list(body)) if body else b''
E       TypeError: sequence item 0: expected a bytes-like object, str found
```

### PR DESCRIPTION
**Pull request by brianbruggeman**
* Fixes #8
* Updates elements within body to use `.encode` which will basically ensure that each element within body is a bytes object rather than a unicode object.  Also bonus - compatible with python 2.

### CODE REVIEW

File: pytest_falcon/plugin.py [hunk 1]
```diff
@@ -141,6 +141,8 @@ def fake_request(self, path, **kwargs):
         environ = create_environ(path, **kwargs)
         resp.environ = environ
         body = self.app(environ, resp)
+        if body:
+            body = [_.encode() if not isinstance(_, bytes) else _ for _ in body]
```
Comment (yohanboniface) [inline pytest_falcon/plugin.py]: Usually, imho `_` as variable name means "something we don't use" (eg. x, `_ = sometuple`), so I'd better use a letter (`p` as `part` or `b` as `body` or `s` as `str`).
Or you can even use a normal name, and split on two lines if we hit the 79 wall:

    body = [part.encode() if not isinstance(part, bytes) else part
            for part in body]

**Review by yohanboniface — CHANGES_REQUESTED** [review]
Thank you :)

Can you also add a very basic unittest to cover the use case your encounter? Thanks :)

File: pytest_falcon/plugin.py (hunk 1 above)
Comment (brianbruggeman) [reply pytest_falcon/plugin.py]: I can modify for a better variable name, but I'd argue that this is exactly the use case for `_`:  it is essentially a variable we do not care about anywhere outside of the list comprehension.

Bikeshedding a little on the variable name:  `part` is not really any better imo than `_`.

Comment (yohanboniface) [reply pytest_falcon/plugin.py]: > I can modify for a better variable name, but I'd argue that this is exactly the use case for _: it is 
> essentially a variable we do not care about anywhere outside of the list comprehension.

That's not my understanding. For me `_` is used for "throwaway" variables, i.e. variable we don't use at all, which is not the case here :)

> Bikeshedding

:)
Use whatever name you want but `_`, which has a special meaning *(by convention)* :)

Thanks!

**Comment by brianbruggeman** [comment]
This should be good to go.  Note I also added sort of a separate review for Pull Request #11.  However, this PR now includes those changes.  More or less the changes are all related to how this package is tested both on Travis-CI and locally.  The old mechanism (`$ pytest`) is still viable.  But the new mechanism (`$ tox`) is more comprehensive.

**Comment by yohanboniface** [comment]
Sorry for my late answer: the PR moved from a simple one line change to a 11 files change, thus it switched from my "easy review" list to my normal todo list, which is much more busy.

The included PR (#11) has nice suggestions, but also some more questionable: adding tox in the workflow (I usually consider a workflow to be better with less tools) or running the CI with non released Python versions (each run takes time, and is using *freely* Travis resources, so I'd say let's be fair).

That said, to be honest, I'm investing less and less time in Falcon (switching my projects to [Roll](https://github.com/pyrates/roll) where I need high speed). So I'd see two options:

- you want to help maintaining pytest-falcon, then I can give you rights, and let you define which workflow is better
- you prefer not to be maintainer, and then I'd say let's move back to the simple change, so we keep the workflow simple for me

Tell me! :)

**Comment by brianbruggeman** [comment]
Hey - thanks for the feedback.

It's one of those catch 22s, right?  Don't change the open source repo too much or it may break something in the ecosystem.  Alternatively, not enough change means that the project stagnates and becomes less and less useful over time.

My roots are in embedded flight controls software for military aircraft, where not being robust means the plane falls out of the sky.  I still like the KISS philosophy for obvious reasons, but I'm not opposed to adding things to make the software more robust - especially if that's in what I consider a tools/process area as opposed to the software application.

My initial change didn't include the test case you wanted because it was pretty complicated to add...but I agree with your initial question/comments/assessment in terms of expansion of the current workflow.

Tox gives the developer the freedom to test their code locally on a matrix of python versions without much developer overhead:  they only need to have the versions installed and only those they're most interested in (e.g. python 2.7.14 and python 3.6.3).  Additionally, Python 3 has at least two "tiers" of changes:  pre 3.4/post 3.4 and pre 3.6/post 3.6.  Given that `pytest-falcon` supports both python 2 and 3, I think that it makes sense to include tox because in my mind, that means _less_ runs on the free travis build waiting to see if a given python version passes.  Having tox setup to run on future versions of python will allow any contributor to see if their changes are future-breaking.  Since Python 3 still feels like it's got some churn, I think that is valuable information.

Everyone has time constraints and I'm no different (job, two young kids, family, etc.).  I want my tools to work so I can focus on the problems I'm running into with my own code.  We use falcon pretty heavily right now, and we're still mostly python 2.  Consequently, I cannot write code at this point that takes advantage of the async features of Python 3.  However, I've been eyeballing the [MagicStack](https://github.com/MagicStack).  Given current falcon benchmarks, I've also considered pypy2.

TLDR; I don't mind helping maintain.  Will you still be involved or are you really looking to pass the whole thing off to someone else?

**Comment by yohanboniface** [comment]
> My roots are in embedded flight controls software

Fun, I do UAVs stuff these days ;)

> Will you still be involved or are you really looking to pass the whole thing off to someone else?

I still have some Falcon stacks running in production, so for now I'll keep an eye. But the direction is moving out to pure python3/asyncio/Roll.
